LanguageFeatures/Records/canonicalization_A01_t02.dart

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion The current specification relies on identical() to decide when to
+/// canonicalize constant object creation expressions. Since identical() is not
+/// useful for records (see below), we update that:
+///
+/// Define two Dart values, a and b, to be structurally equivalent as follows:
+///
+/// If a and b are both records, and they have the same shape, and for each
+/// field f of that shape, the records' values of that field, af and bf are
+/// structurally equivalent, then a and b are structurally equivalent.
+///
+/// If a and b are non-record object references, and they refer to the same
+/// object, then a and b are structurally equivalent. So structural equivalence
+/// agrees with identical() for non-records.
+///
+/// Otherwise a and b are not structurally equivalent.
+///
+/// With that definition, the rules for object and collection canonicalization
+/// is changed from requiring that instance variable, list/set element and map
+/// key/value values are identical() between the instances, to them being
+/// structurally equivalent.
+///
+/// @description Checks that `identical()` is false for records if they are not
+/// structurally equivalent
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=records
+
+import "../../Utils/expect.dart";
+
+main() {
+  if (!isJS) {
+    var r1 = ("a", 1, [42], {"x": 0}, {1, 2, 3});
+    var r2 = ("a", 1.0, [42], {"x": 0}, {1, 2, 3},);
+    Expect.isFalse(identical(r1, r2));
+  }
+
```

Comment:
Added

Reply:
Now we have
```dart
    var r1 = ("a", 1, const [42], const {"x": 0}, const {1, 2, 3});
    var r2 = const ("a", 1.0, [42], {"x": 0}, {1, 2, 3},);
    Expect.isFalse(identical(r1, r2));
```
Do you mean to change it to the below?
```dart
    var r1 = ("a", 1, [42], {"x": 0}, {1, 2, 3});
    var r2 = const ("a", 2, [42], {"x": 0}, {1, 2, 3},);
    var r3 = ("a", 3, const [42], const {"x": 0}, const {1, 2, 3},);
    Expect.isFalse(identical(r1, r2));
    Expect.isFalse(identical(r1, r3));
    Expect.isFalse(identical(r2, r3));
```

Reply:
I wasn't actually thinking about `r1` (I guess it is checked in some other test that records aren't identical when they contain collections obtained from non-constant literals). But the "hardest test" in some sense would be the one where nothing differs among the two records, except for the field which is really different:

```dart
var r1 = const ("a", 1, [42], {"x": 0}, {1, 2, 3},);
var r2 = const ("a", 2, [42], {"x": 0}, {1, 2, 3},);
var r3 = ("a", 2, const [42], const {"x": 0}, const {1, 2, 3},); 
Expect.isFalse(identical(r1, r2));
Expect.isFalse(identical(r1, r3));
```

This would give the tools the smallest possible reason that _forces_ `r1` and `r2` to be distinct representations (so they must be non-identical), and then we have the variant of the same test where `const` is moved into the record.

The reason why we'd want to include the "hardest test" in this sense is that it is the test which is least likely to mask a missing failure (in this case: failure to show that the two records are identical). For instance, if we get "not identical" because `const` has moved then we'd think it works, but we actually got the desired `false` result for a wrong reason.